### ISSUE TITLE
Controling resource deletions

### ISSUE THREAD

**Issue by vavdoshka** [rabbitmq/messaging-topology-operator#922]
**Is your feature request related to a problem? Please describe.**

We are at early stage of adopting the messaging-topology-operator and graduating it's usage to production. Currently removal of a k8s resource triggers the removal of that object in RMQ, we don't feel it safe for our production workloads and would like to control the removal behavior.

**Describe the solution you'd like**

As a flag provided in runtime controlling resource deletion behavior (and/or) on the resource level.

**Describe alternatives you've considered**

Using admission controller to prevent resource removal

**Additional context**

topology-operator can reconcile resources which were requested previously, so it seems the missing piece is logic to ignore delete. 
relevant request on rmq-server side https://github.com/rabbitmq/rabbitmq-server/issues/12772

**Comment by Zerpet** [comment]
A flag to skip deletion of topology objects defeats the point of having an Operator in the first place. The Operator is meant to abstract the management and lifecycle of topology objects in RabbitMQ. Skipping deletion removes a fundamental part of the Operator functionality, and would result in a mismatch between the Kubernetes state and RabbitMQ state; in other words, it would result in objects not present in Kubernetes, and present in RabbitMQ.

**Comment by vavdoshka** [comment]
Introducing a flag to control resource deletion behavior enhances the safety and flexibility of the Messaging Topology Operator without undermining its purpose. Many successful operators, like Crossplane, ArgoCD, Cert-Manager and others provide similar mechanisms to address real-world production needs, ensuring critical resources are not inadvertently deleted. While mismatches between Kubernetes and RabbitMQ states are a concern, they can be mitigated through logging and reconciliation mechanisms. Additionally, we already tested the reconciliation of resources created by the Operator after simulating the orphaned deletion and the existing implementation successfully detected and managed the orphaned resource.

Being used wisely this feature should allow easier and broader adoption of the operator for the both teams being early in their k8s journey and teams managing very sensitive RMQ workloads.

**Comment by ansd** [comment]
@vavdoshka  Why can't you simply disallow deletion of these K8s objects via RBAC (Role Based Access Control)?

**Comment by ansd** [comment]
Also, please reference where these other projects support this feature.

A quick Google search suggests that this is an alpha feature in Crossplane:
https://docs.crossplane.io/latest/concepts/usages/

It's an `open` issue in ArgoCD:
https://github.com/argoproj/argo-cd/issues/7301
https://github.com/argoproj/argo-cd/issues/16063

**Comment by Zerpet** [comment]
> Introducing a flag to control resource deletion behavior enhances the safety and flexibility of the Messaging Topology Operator without undermining its purpose.

I disagree. The mission of the Operator is to reconcile Kubernetes state into RabbitMQ. This flag is completely opposed to that mission.

> ensuring critical resources are not inadvertently deleted

How can you inadvertently delete a resource when a deletion requires a explicit `kubectl delete` action, or an API DELETE request? It sounds like we are trying to fix human carelessness, and inherently unfixable problem.

> Being used wisely this feature should allow easier and broader adoption of the operator for the both teams being early in their k8s journey and teams managing very sensitive RMQ workloads.

Given that this is the first time this feature has been requested in the ~4 years of existence of this Operator, I'm highly skeptical that it will help to spread adoption.

**Comment by Zerpet** [comment]
> I propose to let users decide whether to optionally retain the resources that might contain a state influenced by factors outside the operator's responsibility or concern, i.e., queues or streams with data. Optional retain behavior, similar to what K8s implements for PVC Reclaim Policy.

This sounds perfectly reasonable. I'm compelled by the reclaim policy analogy. I agree it makes sense to tweak the "reclaim" behaviour for resources that contain data, in other words: queues, streams and vhosts. A vhost does not contain data per-se, but its sub-resources (queues) do. Other resources that do not contain data e.g. policies, I'm not sure about implementing a reclaim policy.

> We are trying to understand if maintainers would be opened to such contribution at current moment in order to plan accordingly the actions on our side. 

I'm happy to go forth with a reclaim policy feature for `Queue` and `VHost` objects.

Any thoughts @mkuratczyk @ansd ?

**Comment by michaelklishin** [comment]
Policies, runtime parameters, virtual host limits do not need a reclaim policy. It is extremely rare to see those things consuming a non-trivial amount of either memory or disk space.

Shovels and Federation upstreams, both based on runtime parameters, might need it because a shovel or a federation link can be configured to have a very high prefetch.

Users and permissions need to be treated differently because their rotation is typically not driven by how much memory or compute resources they take but rather by the credential rotation policy, which can vary greatly from deployment to deployment. So I'd leave those two resources out, at first or until/if something more suitable for them is designed and proposed.

**Comment by michaelklishin** [comment]
> I'm happy to go forth with a reclaim policy feature for Queue and VHost objects

I assume "queues' implicitly includes streams because they are considered a queue type, not a separate entity entirely, even though in most ways they are.
